src/runtime/cuda/l2_cache_flush.cc

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+// Acknowledgement: l2flush struct in nvbench project.
+// Reference:
+// https://github.com/NVIDIA/nvbench/blob/1a13a2e724b8aa8aee27649ac6878babb63862a6/nvbench/detail/l2flush.cuh
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <dmlc/thread_local.h>
+#include <tvm/runtime/device_api.h>
+#include <tvm/runtime/registry.h>
+
+#include "cuda_common.h"
+
+namespace tvm {
+
+namespace runtime {
+
+class L2Flush {
+ public:
+  L2Flush() : initialized_(false), l2_size_(0), l2_buffer_(nullptr) {}
+
+  ~L2Flush() {
+    if (l2_size_ > 0) {
+      CUDA_CALL(cudaFree(l2_buffer_));
+    }
+  }
+
+  void Flush() {
+    if (!initialized_) {
```

Comment:
we expect that after cudaMalloc the variable l2_buffer_ will not be nullptr? then can we use this instead of initialized_?

Reply:
1. Yes
2. `initialized` was used to determine whether it was the first run. We don't want to call this malloc function if `Flush` was never triggered so we don't put them under initialization function.

Reply:
I think that @Icemist wanted to propose removing `initialized_` variable and check `l2_buffer_` instead:
```c++
  void Flush() {
    if (l2_buffer_ == nullptr) {
      // ...
```

With the same idea, I have asked my question about `l2_size_` variable in the destructor.

Reply:
Per [CUDA runtime documentation](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html): 
>[cudaDevAttrL2CacheSize](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gg49e2f8c2c0bd6fe264f2fc970912e5cd1b0342682d15910022ba3f383a851ad7): Size of L2 cache in bytes. 0 if the device doesn't have L2 cache.

`l2_size_` would be set to 0 if the device does not have `l2`, in which case `l2_buffer_` is still a nullptr I suppose.

Reply:
Thank you for the clarification! Yes, in this case I think `initialized_` is the better solution.